import DeviseService from "../../services/devise"
import { Router, Request, Response, NextFunction } from "express"
import { Container } from "typedi"

const create = async (req: Request, res: Response, next: NextFunction) => {
    const logger = Container.get("logger")
    const{user_code} = req.headers

    logger.debug("Calling Create code endpoint")
    try {
        const deviseServiceInstance = Container.get(DeviseService)
        const devise = await deviseServiceInstance.create({...req.body, created_by:user_code,created_ip_adr: req.headers.origin, last_modified_by:user_code,last_modified_ip_adr: req.headers.origin})
        return res
            .status(201)
            .json({ message: "created succesfully", data:  devise })
    } catch (e) {
        //#
        logger.error("ðŸ”¥ error: %o", e)
        return next(e)
    }
}

const findOne = async (req: Request, res: Response, next: NextFunction) => {
    const logger = Container.get("logger")
    logger.debug("Calling find one  code endpoint")
    try {
        const deviseServiceInstance = Container.get(DeviseService)
        const {id} = req.params
        const devise = await deviseServiceInstance.findOne({id})
        return res
            .status(200)
            .json({ message: "fetched succesfully", data: devise  })
    } catch (e) {
        logger.error("ðŸ”¥ error: %o", e)
        return next(e)
    }
}

const findAll = async (req: Request, res: Response, next: NextFunction) => {
    const logger = Container.get("logger")
    logger.debug("Calling find all code endpoint")
    try {
        const deviseServiceInstance = Container.get(DeviseService)
        const devise = await deviseServiceInstance.find({})
        return res
            .status(200)
            .json({ message: "fetched succesfully", data: devise })
    } catch (e) {
        logger.error("ðŸ”¥ error: %o", e)
        return next(e)
    }
}

const findBy = async (req: Request, res: Response, next: NextFunction) => {
    const logger = Container.get("logger")
    logger.debug("Calling find by  all code endpoint")
    try {
        const deviseServiceInstance = Container.get(DeviseService)
        const devise = await deviseServiceInstance.findOne({...req.body})
        return res
            .status(200)
            .json({ message: "fetched succesfully", data: devise })
    } catch (e) {
        logger.error("ðŸ”¥ error: %o", e)
        return next(e)
    }
}

const update = async (req: Request, res: Response, next: NextFunction) => {
    const logger = Container.get("logger")
    const{user_code} = req.headers

    logger.debug("Calling update one  code endpoint")
    try {
        const deviseServiceInstance = Container.get(DeviseService)
        const {id} = req.params
        const devise = await deviseServiceInstance.update({...req.body, last_modified_by:user_code,last_modified_ip_adr: req.headers.origin},{id})
        return res
            .status(200)
            .json({ message: "fetched succesfully", data: devise  })
    } catch (e) {
        logger.error("ðŸ”¥ error: %o", e)
        return next(e)
    }
}

const deleteOne = async (req: Request, res: Response, next: NextFunction) => {
    const logger = Container.get("logger")
    logger.debug("Calling update one  code endpoint")
    try {
        const deviseServiceInstance = Container.get(DeviseService)
        const {id} = req.params
        const devise = await deviseServiceInstance.delete({id})
        return res
            .status(200)
            .json({ message: "deleted succesfully", data: id  })
    } catch (e) {
        logger.error("ðŸ”¥ error: %o", e)
        return next(e)
    }
}

export default {
    create,
    findOne,
    findAll,
    findBy,
    update,
    deleteOne
}

